KAFKA-9218: MirrorMaker 2 can fail to create topics

edoardocomar opened the pull request:
When the scheduled refreshTopicPartitions runs, check existing topics in both source and target clusters in order to compute topic partitions to be created on target.

If a temporary failure to create the target topic is encountered (e.g. insufficient number of brokers), on the next refresh the target topic creation will be re-attempted.

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@uk.ibm.com>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

edoardocomar committed:
KAFKA-9218: MirrorMaker 2 can fail to create topics

Check topics in both source and target clusters to compute topic
partitions to create on target.

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@uk.ibm.com>

edoardocomar commented:
test failure on JDK 11 and Scala 2.13 unrelated to this PR

edoardocomar commented:
@ryannedolan your comments welcome

mimaison commented:
@bbejeck @omkreddy can you take a look?

This PR ensures that when MM2 fails to create a topic, it retries the next time around.
I think the code could be improved further and only attempt CreatePartitions when there's something to do but we cna do that in a follow up PR.

ryannedolan commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -162,57 +162,70 @@ public String version() {
-        List<TopicPartition> topicPartitions = findTopicPartitions();
+        knownSourceTopicPartitions = findSourceTopicPartitions();
+        knownTargetTopicPartitions = findExistingTargetTopicPartitions();
+        List<TopicPartition> targetTopicPartitionsUpstream = knownTargetTopicPartitions.stream()
```
I don't have a better suggestion, but targetTopicPartitionsUpstream is a kinda confusing name.

edoardocomar replied:
`upstreamTargetTopicPartitions` is certainly a *much* better name :-)

ryannedolan commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -243,18 +256,29 @@ private void createOffsetSyncsTopic() {
+        Map<String, Long> partitionCounts = knownSourceTopicPartitions.stream()
             .collect(Collectors.groupingBy(x -> x.topic(), Collectors.counting())).entrySet().stream()
             .collect(Collectors.toMap(x -> formatRemoteTopic(x.getKey()), x -> x.getValue()));
+        Set<String> knownTargetTopics = knownTargetTopicPartitions.stream()
```
This logic appears twice. Can we extract to a method?

ryannedolan reviewed (approved):
Thanks for finding and fixing this issue! The commit message could better explain what is being fixed and how. A couple other nits, otherwise lgtm.

ryannedolan commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -112,4 +128,38 @@ public void testConfigPropertyFiltering() {
     }
+
+    @Test
+    public void testRefreshTopicPartitions() throws Exception {
```
Mockito is mysterious to me -- someone else should review this please.

edoardocomar committed:
Addressing comments

The original commit comment could be reworded as:

When the scheduled refreshTopicPartitions runs, check existing topics in
both source and target clusters in order to compute topic partitions to
be created on target.

If a temporary failure to create the target topic is encountered (e.g.
insufficient number of brokers), on the next refresh the target topic
creation will be re-attempted.


Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@uk.ibm.com>

edoardocomar commented:
Thanks @ryannedolan I've pushed some cosmetic changes to address your comments.

ryannedolan commented:
Please rebase, otherwise lgtm.

edoardocomar commented:
Thanks @ryannedolan , rebased 

@omkreddy could you review please ?

omkreddy reviewed (approved):
@edoardocomar Thanks for the PR. LGTM.
@mimaison feel free to merge.
